### PR TITLE
Make command line parsing more robust

### PR DESCRIPTION
**Pull request by SimonTeixidor**
This patch adds command line switches and prints a descriptive message if you messed something up. Further, it adds the ability to read password during execution rather than as a parameter.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -1,27 +1,65 @@
...
+use getopts::Options;
+use rpassword::read_password;
+
 use librespot::session::{Config, Session};
 use librespot::util::version::version_string;
 use librespot::player::Player;
 use librespot::spirc::SpircManager;
 
+fn usage(program: &str, opts: &Options) -> String {
+    let brief = format!("Usage: {} FILE [options]", program);
```
Comment (plietar) [inline src/main.rs]: The `FILE` seems to be a leftover from getopts's doc

File: src/main.rs [hunk 2]
```diff
@@ -1,27 +1,65 @@
...
+    format!("{}", opts.usage(&brief)).to_owned()
```
Comment (plietar) [inline src/main.rs]: The `to_owned` isn't needed.

File: src/main.rs [hunk 3]
```diff
@@ -1,27 +1,65 @@
...
-    let name = args.next().unwrap();
+    let args: Vec<String> = std::env::args().collect();
+    let program = args[0].clone();
+
+    let mut opts = Options::new();
+    opts.reqopt("a", "", "Path to a spotify appkey", "APPKEY");
+    opts.reqopt("u", "", "Username to sign in with", "USERNAME");
+    opts.optopt("p", "", "Password (optional)", "PASSWORD");
+    opts.reqopt("c", "", "Path to a directory where files will be cached.", "CACHE");
+    opts.reqopt("d", "", "Device name", "DEVICE");
```
Comment (plietar) [inline src/main.rs]: Make that `-n`/`--name`

File: src/main.rs [hunk 4]
```diff
@@ -1,27 +1,65 @@
...
-    let mut appkey_file = File::open(Path::new(&args.next().unwrap())).unwrap();
-    let username = args.next().unwrap();
-    let password = args.next().unwrap();
-    let cache_location = args.next().unwrap();
-    let name = args.next().unwrap();
+    let args: Vec<String> = std::env::args().collect();
+    let program = args[0].clone();
+
+    let mut opts = Options::new();
+    opts.reqopt("a", "", "Path to a spotify appkey", "APPKEY");
```
Comment (plietar) [inline src/main.rs]: Could you add the long versions as well

File: src/main.rs [hunk 5]
```diff
@@ -1,27 +1,65 @@
...
+
+    let mut appkey_file = File::open(
+                                    Path::new(&*matches.opt_str("a").unwrap())
+                                ).expect("Could not open app key.");
+
+    let username = matches.opt_str("u").unwrap();
+    let cache_location = matches.opt_str("c").unwrap();
+    let name = matches.opt_str("d").unwrap();
+
+    let password = match matches.opt_str("p") {
```
Comment (plietar) [inline src/main.rs]: You can use `Result::unwrap_or_else` for this.

File: README.md [hunk 6]
```diff
@@ -25,7 +25,7 @@ cargo build
 A sample program implementing a headless Spotify Connect receiver is provided.
 Once you've built *librespot*, run it using :
 ```shell
-target/debug/main APPKEY USERNAME PASSWORD CACHEDIR DEVICENAME
+target/debug/main -a APPKEY -u USERNAME -c CACHEDIR -d DEVICENAME
 ```
 where `APPKEY` is the path to a Spotify application key file, `USERNAME` and
 `PASSWORD` are your Spotify credentials, `CACHEDIR` is the path to directory
```
Comment (plietar) [inline README.md]: The `PASSWORD` part should be removed.

**Comment by plietar** [comment]
Thanks a lot. I've made some comments, if you don't mind updating the PR please.

**Comment by SimonTeixidor** [comment]
I've pushed some updates. Tell me if you're satisfied and I'll rebase it to one commit.

File: src/main.rs [hunk 7]
```diff
@@ -1,27 +1,63 @@
...
-    let name = args.next().unwrap();
+    let args: Vec<String> = std::env::args().collect();
+    let program = args[0].clone();
+
+    let mut opts = Options::new();
+    opts.reqopt("a", "appkey", "Path to a spotify appkey", "APPKEY");
+    opts.reqopt("u", "username", "Username to sign in with", "USERNAME");
+    opts.optopt("p", "password", "Password (optional)", "PASSWORD");
+    opts.reqopt("c", "cache", "Path to a directory where files will be cached.", "CACHE");
+    opts.reqopt("n", "name", "Device name", "DEVICE");
```
Comment (plietar) [inline src/main.rs]: That should be `NAME`

File: src/main.rs [hunk 8]
```diff
@@ -1,27 +1,63 @@
...
+    let mut appkey_file = File::open(
+                                    Path::new(&*matches.opt_str("a").unwrap())
+                                ).expect("Could not open app key.");
+
+    let username = matches.opt_str("u").unwrap();
+    let cache_location = matches.opt_str("c").unwrap();
+    let name = matches.opt_str("n").unwrap();
+
+    let password = matches.opt_str("p").unwrap_or_else(||
+        {
```
Comment (plietar) [inline src/main.rs]: Small nit, but could you put the brace on the first line, and reduce the indentation of the block by a tab

**Comment by SimonTeixidor** [comment]
Done and done!

**Comment by plietar** [comment]
LGTM, if you could just squash them and I can merge

**Comment by SimonTeixidor** [comment]
Now they're squashed. :)

**Comment by plietar** [comment]
Thanks

**Pull request merged**